### PR TITLE
Click Dragging A Fluid To A Reagent Container Now Attempts To Transfer Fluid To That Container

### PR DESCRIPTION
**Pull request by Mister-Moriarty**
[Player Actions] [Feature] [QoL]

## About the PR:
Similar to dragging one reagent container to another, dragging a fluid to a container will now attempt to transfer some of the fluid to that container, the volume transferred currently being at minimum the empty volume remaining in the container, and at maximum the total amount of reagents per tile covered by the fluid puddle.
To facilitate the above, the `mousedrop_refill` list (and the many snowflake checks used in place of it) was refactored into a bitflag system.

## Why's this needed?
Quality of life; attempting to transfer reagents from a fluid puddle to a container requires both patience and either using two containers to slowly transfer reagents or clicking inbetween fluid updates to dump then swiftly pick up the reagents again.

## Changelog:
```changelog
(u)Mr. Moriarty
(+)Click dragging a fluid puddle to a reagent container now attempts to transfer fluid to that container.
```

### CODE REVIEW

File: code/modules/fluids/fluid_core.dm [hunk 1]
```diff
@@ -214,6 +214,28 @@ var/mutable_appearance/fluid_ma
 		var/turf/T = src.loc
 		T.Attackhand(user)
 
+	mouse_drop(obj/over_object, src_location, over_location)
+		if (istype(src, /obj/fluid/airborne) || !istype(over_object, /obj/item/reagent_containers/glass) && !istype(over_object, /obj/item/reagent_containers/food/drinks) && !istype(over_object, /obj/reagent_dispensers) && !istype(over_object, /obj/item/spraybottle) && !istype(over_object, /obj/machinery/plantpot) && !istype(over_object, /obj/mopbucket) && !istype(over_object, /obj/item/reagent_containers/mender) && !istype(over_object, /obj/item/tank/jetpack/backtank) && !istype(over_object, /obj/machinery/bathtub))
```
Comment (pgmzeta) [inline code/modules/fluids/fluid_core.dm:218]: any way to use `mousedrop_refill` here?

**Comment by BatElite** [comment]
Does the max amount match what you'd get clicking on a puddle with an empty container atm? I'm not that familiar.

**Comment by pgmzeta** [comment]
floor ants/spiders are `reagent_dispensers`, technically:
`You fill the space ants with 12.5 units of the ants.x2`

You can also fill up a water/compost/fuel tank very quickly:
```You fill the high-capacity watertank with 276.389 units of the water.
You fill the high-capacity watertank with 268.711 units of the water.
You fill the high-capacity watertank with 261.247 units of the water.
You fill the high-capacity watertank with 253.99 units of the water.```


Maybe this should be restricted to glasses/beakers/buckets/mopbuckets?

**Comment by Mister-Moriarty** [comment]
> floor ants/spiders are `reagent_dispensers`, technically: `You fill the space ants with 12.5 units of the ants.x2`
> 
> You can also fill up a water/compost/fuel tank very quickly:
> 
> ```
> You fill the high-capacity watertank with 268.711 units of the water.
> You fill the high-capacity watertank with 261.247 units of the water.
> You fill the high-capacity watertank with 253.99 units of the water.
> ```
> 
> Maybe this should be restricted to glasses/beakers/buckets/mopbuckets?

I believe the simplest and most intuitive resolution would be to only allow containers that are in `usr.contents` to be able to be filled with a click-drag from a fluid puddle, as to avoid click dragging to reagent tanks or ants/spiders on the floors.

Regarding the volume of the reagent tanks filling rather rapidly, the code takes the lower of the capacity remaining in the reagent container and the amount of fluid per tile, so for 268.711u to be transferred to a tank, there would have to be 268.711u in a single tile.

File: code/modules/fluids/fluid_core.dm (hunk 1 above)
Comment (Mister-Moriarty) [reply code/modules/fluids/fluid_core.dm:218]: The `istype` checks were pinched from the `mouse_drop` code for `/obj/machinery/bathtub`, and while using the `mousedrop_refill` list would be far cleanlier than a list of `istype` checks, I'm rather unsure as to how to correctly implement it.

**Comment by Mister-Moriarty** [comment]
> Does the max amount match what you'd get clicking on a puddle with an empty container atm? I'm not that familiar.

It should, yes. The difference being that you may continue to click-drag to continue filling the reagent container.

**Comment by Garash2k** [comment]
Is there a concern where this obsoletes mops? Is it ok if scooping with a bucket and splashing it at a wall faster than mopping up the puddle?

**Pull request closed without merging**